File: app/src/main/java/closer/vlllage/com/closer/handler/data/ApiHandler.java
package closer.vlllage.com.closer.handler.data;

import com.google.android.gms.maps.model.LatLng;

import java.io.IOException;
import java.io.InputStream;
import java.util.Date;
import java.util.List;

import javax.annotation.Nullable;

import closer.vlllage.com.closer.api.ApiService;
import closer.vlllage.com.closer.api.models.CreateResult;
import closer.vlllage.com.closer.api.models.EventResult;
import closer.vlllage.com.closer.api.models.GroupActionResult;
import closer.vlllage.com.closer.api.models.GroupMemberResult;
import closer.vlllage.com.closer.api.models.GroupMessageResult;
import closer.vlllage.com.closer.api.models.GroupResult;
import closer.vlllage.com.closer.api.models.PhoneResult;
import closer.vlllage.com.closer.api.models.ReactionResult;
import closer.vlllage.com.closer.api.models.StateResult;
import closer.vlllage.com.closer.api.models.SuccessResult;
import closer.vlllage.com.closer.api.models.SuggestionResult;
import closer.vlllage.com.closer.handler.helpers.DateFormatter;
import closer.vlllage.com.closer.handler.helpers.HttpEncode;
import closer.vlllage.com.closer.handler.helpers.LatLngStr;
import closer.vlllage.com.closer.handler.helpers.Val;
import closer.vlllage.com.closer.pool.PoolMember;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import okhttp3.internal.Util;
import okio.BufferedSink;
import okio.Okio;
import okio.Source;

public class ApiHandler extends PoolMember {

    private final ApiService api = new ApiService();

    public void setAuthorization(String auth) {
        api.setAuthorization(auth);
    }

    public Single<List<PhoneResult>> getPhonesNear(LatLng latLng) {
        return uiThread(api.getBackend().getPhonesNear($(LatLngStr.class).from(latLng)));
    }

    public Single<List<SuggestionResult>> getSuggestionsNear(LatLng latLng) {
        return uiThread(api.getBackend().getSuggestionsNear($(LatLngStr.class).from(latLng)));
    }

    public Single<CreateResult> addSuggestion(String name, LatLng latLng) {
        return uiThread(api.getBackend().addSuggestion(name, $(LatLngStr.class).from(latLng)));
    }

    public Single<CreateResult> updatePhone(String latLng, String name, String status, Boolean active, String deviceToken) {
        return uiThread(api.getBackend().phoneUpdate(latLng, name, status, active, deviceToken));
    }

    public Single<CreateResult> updatePhonePhoto(String photoUrl) {
        return uiThread(api.getBackend().phoneUpdatePhoto(photoUrl));
    }

    public Single<CreateResult> updatePhonePrivateMode(boolean privateMode) {
        return uiThread(api.getBackend().updatePhonePrivateMode(privateMode));
    }

    public Single<PhoneResult> phone() {
        return uiThread(api.getBackend().phone());
    }

    public Single<List<PhoneResult>> searchPhonesNear(LatLng latLng, String query) {
        return uiThread(api.getBackend().searchPhonesNear($(LatLngStr.class).from(latLng), query));
    }

    public Single<SuccessResult> sendMessage(String phone, String message) {
        return uiThread(api.getBackend().sendMessage(phone, message));
    }

    public Single<Boolean> isVerified() {
        return uiThread(api.getBackend().getIsVerified())
                .map(verifiedResult -> verifiedResult.verified);
    }

    public Single<List<GroupMessageResult>> myMessages(LatLng latLng) {
        return uiThread(api.getBackend().myMessages($(LatLngStr.class).from(latLng)));
    }

    public Single<StateResult> myGroups(LatLng latLng) {
        return uiThread(api.getBackend().myGroups($(LatLngStr.class).from(latLng)));
    }

    public Single<GroupResult> getGroup(String groupId) {
        return uiThread(api.getBackend().getGroup(groupId));
    }

    public Single<SuccessResult> setPhoneNumber(String phoneNumber) {
        return uiThread(api.getBackend().setPhoneNumber(phoneNumber));
    }

    public Single<SuccessResult> sendVerificationCode(String verificationCode) {
        return uiThread(api.getBackend().sendVerificationCode(verificationCode));
    }

    public Single<CreateResult> sendGroupMessage(String groupId, String text, String attachment) {
        return uiThread(api.getBackend().sendGroupMessage(groupId, text, attachment));
    }

    public Single<SuccessResult> reactToMessage(String messageId, String reaction, boolean removeReaction) {
        return uiThread(api.getBackend().reactToMessage(messageId, reaction, removeReaction));
    }

    public Single<List<ReactionResult>> groupMessageReactions(String messageId) {
        return uiThread(api.getBackend().groupMessageReactions(messageId));
    }

    public Single<CreateResult> sendAreaMessage(LatLng latLng, String text, String attachment) {
        return uiThread(api.getBackend().sendAreaMessage($(LatLngStr.class).from(latLng), text, attachment));
    }

    public Single<CreateResult> createGroup(String groupName) {
        return uiThread(api.getBackend().createGroup(groupName));
    }

    public Single<CreateResult> createPublicGroup(String groupName, String about, LatLng latLng) {
        return uiThread(api.getBackend().createPublicGroup(groupName, about, $(LatLngStr.class).from(latLng), true));
    }

    public Single<CreateResult> createPhysicalGroup(LatLng latLng) {
        return uiThread(api.getBackend().createPhysicalGroup($(LatLngStr.class).from(latLng), true));
    }

    public Single<SuccessResult> convertToHub(String groupId, String name) {
        return uiThread(api.getBackend().convertToHub(groupId, name, true));
    }

    public Single<SuccessResult> inviteToGroup(String groupId, String name, String phoneNumber) {
        return uiThread(api.getBackend().inviteToGroup(groupId, name, phoneNumber));
    }

    public Single<SuccessResult> inviteToGroup(String groupId, String phoneId) {
        return uiThread(api.getBackend().inviteToGroup(groupId, phoneId));
    }

    public Single<SuccessResult> leaveGroup(String groupId) {
        return uiThread(api.getBackend().leaveGroup(groupId, true));
    }

    public Single<SuccessResult> setGroupPhoto(String groupId, String photo) {
        return uiThread(api.getBackend().setGroupPhoto(groupId, photo));
    }

    public Single<SuccessResult> setGroupAbout(String groupId, String about) {
        return uiThread(api.getBackend().setGroupAbout(groupId, about));
    }

    public Single<SuccessResult> removeGroupAction(String groupActionId) {
        return uiThread(api.getBackend().removeGroupAction(groupActionId));
    }

    public Single<CreateResult> createGroupAction(String groupId, String name, String intent) {
        return uiThread(api.getBackend().createGroupAction(groupId, name, intent));
    }

    public Single<List<GroupActionResult>> getGroupActions(String groupId) {
        return uiThread(api.getBackend().getGroupActions(groupId));
    }

    public Single<List<GroupActionResult>> getGroupActions(LatLng latLng) {
        return uiThread(api.getBackend().getGroupActionsNearGeo($(LatLngStr.class).from(latLng)));
    }

    public Single<List<GroupMessageResult>> getGroupMessages(String groupId) {
        return uiThread(api.getBackend().getGroupMessages(groupId));
    }

    public Single<GroupMessageResult> getGroupMessage(String groupMessageId) {
        return uiThread(api.getBackend().getGroupMessage(groupMessageId));
    }

    public Single<SuccessResult> setGroupActionPhoto(String groupActionId, String photo) {
        return uiThread(api.getBackend().setGroupActionPhoto(groupActionId, photo));
    }

    public Single<List<GroupMemberResult>> getAllGroupMember() {
        return uiThread(api.getBackend().getAllGroupMembers());
    }

    public Single<GroupMemberResult> getGroupMember(String groupId) {
        return uiThread(api.getBackend().getGroupMember(groupId));
    }

    public Single<CreateResult> updateGroupMember(String groupId, boolean muted, boolean subscribed) {
        return uiThread(api.getBackend().updateGroupMember(groupId, muted, subscribed));
    }

    public Single<SuccessResult> cancelInvite(String groupId, String groupInviteId) {
        return uiThread(api.getBackend().cancelInvite(groupId, groupInviteId));
    }

    public Single<CreateResult> createEvent(String name, String about, boolean isPublic, LatLng latLng, Date startsAt, Date endsAt) {
        return uiThread(api.getBackend().createEvent(name, about, isPublic, $(LatLngStr.class).from(latLng), $(HttpEncode.class).encode($(DateFormatter.class).format(startsAt)), $(HttpEncode.class).encode($(DateFormatter.class).format(endsAt))));
    }

    public Single<List<EventResult>> getEvents(LatLng latLng) {
        return uiThread(api.getBackend().getEvents($(LatLngStr.class).from(latLng)));
    }

    public Single<EventResult> getEvent(String eventId) {
        return uiThread(api.getBackend().getEvent(eventId));
    }

    public Single<SuccessResult> cancelEvent(String eventId) {
        return uiThread(api.getBackend().cancelEvent(eventId, true));
    }

    public Single<List<GroupResult>> getPhysicalGroups(LatLng latLng) {
        return uiThread(api.getBackend().getPhysicalGroups($(LatLngStr.class).from(latLng), "physical"));
    }

    public Single<String> uploadPhoto(InputStream photo) {
        RequestBody body = new RequestBody() {
            @Nullable
            @Override
            public MediaType contentType() {
                return MediaType.parse("image/*");
            }

            @Override
            public long contentLength() {
                try {
                    return photo.available();
                } catch (IOException e) {
                    e.printStackTrace();
                    return 0;
                }
            }

            @Override
            public void writeTo(BufferedSink sink) {
                Source source = Okio.source(photo);
                try {
                    sink.writeAll(source);
                } catch (IOException e) {
                    e.printStackTrace();
                } finally {
                    Util.closeQuietly(source);
                }
            }
        };
        MultipartBody.Part part = MultipartBody.Part.createFormData("photo", "closer-photo", body);
        String id = $(Val.class).rndId();
        return uiThread(api.getPhotoUploadBackend().uploadPhoto(id, part))
                .map(responseBody -> id);
    }

    public Single<String> privacy() {
        return uiThread(api.getContentBackend().privacy()).map(ResponseBody::string);
    }

    private <T> Single<T> uiThread(Single<T> observable) {
        return observable.observeOn(AndroidSchedulers.mainThread());
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/data/ApiHandler.kt
package closer.vlllage.com.closer.handler.data

import closer.vlllage.com.closer.api.ApiService
import closer.vlllage.com.closer.api.Backend
import closer.vlllage.com.closer.api.models.QuestProgressResult
import closer.vlllage.com.closer.api.models.QuestResult
import closer.vlllage.com.closer.api.models.SuccessResult
import closer.vlllage.com.closer.handler.event.EventReminder
import closer.vlllage.com.closer.handler.helpers.*
import closer.vlllage.com.closer.store.models.QuestFlow
import closer.vlllage.com.closer.store.models.QuestProgressFlow
import com.google.android.gms.maps.model.LatLng
import com.queatz.on.On
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody
import okhttp3.internal.Util
import okio.BufferedSink
import okio.Okio
import retrofit2.HttpException
import java.io.IOException
import java.io.InputStream
import java.util.*

class ApiHandler constructor(private val on: On) {

    private val api = ApiService()

    fun isVerified() = api { isVerified() }.map { it.verified }

    fun allGroupMember() = api { allGroupMembers() }

    fun setAuthorization(auth: String) {
        api.authorization = auth
    }

    fun terms(phoneId: String, good: Boolean) = api { terms(phoneId, good) }

    fun call(phoneId: String, event: String, data: String) = api { call(phoneId, event, data) }

    fun getPhonesNear(latLng: LatLng) = api { getPhonesNear(on<LatLngStr>().from(latLng)) }

    fun getSuggestionsNear(latLng: LatLng) = api { getSuggestionsNear(on<LatLngStr>().from(latLng)) }

    fun addSuggestion(name: String, latLng: LatLng) = api { addSuggestion(name, on<LatLngStr>().from(latLng)) }

    fun getStoriesNear(latLng: LatLng) = api { getStoriesNear(on<LatLngStr>().from(latLng)) }

    fun addStory(text: String?, photo: String?, latLng: LatLng) = api { addStory(text, photo, on<LatLngStr>().from(latLng)) }

    fun storyViewed(storyId: String) = api { storyViewed(storyId) }

    fun getStory(storyId: String) = api { getStory(storyId) }

    fun updatePhone(latLng: String? = null,
                    name: String? = null,
                    status: String? = null,
                    active: Boolean? = null,
                    deviceToken: String? = null,
                    introduction: String? = null,
                    offtime: String? = null,
                    occupation: String? = null,
                    history: String? = null) = api { phoneUpdate(latLng, name, status, active, deviceToken, introduction, offtime, occupation, history) }

    fun updatePhonePhoto(photoUrl: String) = api { phoneUpdatePhoto(photoUrl) }

    fun updatePhonePrivateMode(privateMode: Boolean) = api { updatePhonePrivateMode(privateMode) }

    fun phone() = api { phone() }

    fun searchPhonesNear(latLng: LatLng, query: String) = api { searchPhonesNear(on<LatLngStr>().from(latLng), query) }

    fun getPhone(phoneId: String) = api { getPhone(phoneId) }

    fun sendMessage(phone: String, message: String) = api { sendMessage(phone, message) }

    fun myMessages(latLng: LatLng) = api { myMessages(on<LatLngStr>().from(latLng)) }

    fun myGroups(latLng: LatLng) = api { myGroups(on<LatLngStr>().from(latLng)) }

    fun myEvents() = api { myEvents() }

    fun getGroup(groupId: String) = api { getGroup(groupId) }

    fun getDirectGroup(phoneId: String) = api { getDirectGroup(phoneId) }

    fun setPhoneNumber(phoneNumber: String) = api { setPhoneNumber(phoneNumber) }

    fun sendVerificationCode(verificationCode: String) = api { sendVerificationCode(verificationCode) }

    fun addGoal(goalName: String, remove: Boolean? = null) = api { addGoal(goalName, remove) }

    fun addLifestyle(lifestyleName: String, remove: Boolean? = null) = api { addLifestyle(lifestyleName, remove) }

    fun phonesForGoal(goalName: String) = api { phonesForGoal(goalName) }

    fun phonesForLifestyle(lifestyleName: String) = api { phonesForLifestyle(lifestyleName) }

    fun getNewPhonesToMeetNear(latLng: LatLng) = api { getNewPhonesToMeetNear(on<LatLngStr>().from(latLng)) }

    fun setMeet(phoneId: String, meet: Boolean) = api { setMeet(phoneId, meet) }

    fun sendGroupMessage(groupId: String, text: String?, attachment: String?) = api { sendGroupMessage(groupId, text, attachment) }

    fun deleteGroupMessage(messageId: String) = api { deleteGroupMessage(messageId, true) }

    fun reactToMessage(messageId: String, reaction: String, removeReaction: Boolean) = api { reactToMessage(messageId, reaction, removeReaction) }

    fun groupMessageReactions(messageId: String) = api { groupMessageReactions(messageId) }

    fun createGroup(groupName: String) = api { createGroup(groupName) }

    fun createPublicGroup(groupName: String, about: String, latLng: LatLng) = api { createPublicGroup(groupName, about, on<LatLngStr>().from(latLng), true) }

    fun createPhysicalGroup(name: String, about: String, isPublic: Boolean, latLng: LatLng, hub: Boolean) = api { createPhysicalGroup(name, about, on<LatLngStr>().from(latLng), isPublic, true, hub) }

    fun convertToHub(groupId: String, name: String) = api { convertToHub(groupId, name, true) }

    fun inviteToGroup(groupId: String, name: String, phoneNumber: String) = api { inviteToGroup(groupId, name, phoneNumber) }

    fun inviteToGroup(groupId: String, phoneId: String) = api { inviteToGroup(groupId, phoneId) }

    fun leaveGroup(groupId: String) = api { leaveGroup(groupId, true) }

    fun updateGroupStatus(groupId: String, status: String) = api { updateGroupStatus(groupId, status) }

    fun updateGroupPhoto(groupId: String, photo: String) = api { updateGroupPhoto(groupId, photo) }

    fun setGroupPhoto(groupId: String, photo: String) = api { setGroupPhoto(groupId, photo) }

    fun setGroupAbout(groupId: String, about: String) = api { setGroupAbout(groupId, about) }

    fun removeGroupAction(groupActionId: String) = api { removeGroupAction(groupActionId) }

    fun createGroupAction(groupId: String, name: String, intent: String) = api { createGroupAction(groupId, name, intent) }

    fun updateGroupActionFlow(groupActionId: String, flow: String) = api { updateGroupActionFlow(groupActionId, flow) }

    fun updateGroupActionAbout(groupActionId: String, about: String) = api { updateGroupActionAbout(groupActionId, about) }

    fun usedGroupAction(groupActionId: String) = api { usedGroupAction(groupActionId, true) }

    fun getGroupAction(groupActionId: String) = api { getGroupAction(groupActionId) }

    fun getGroupActions(groupId: String) = api { getGroupActions(groupId) }

    fun getGroupActions(latLng: LatLng) = api { getGroupActionsNearGeo(on<LatLngStr>().from(latLng)) }

    fun getPins(groupId: String) = api { getPins(groupId) }

    fun getContacts(groupId: String) = api { getContacts(groupId) }

    fun addPin(messageId: String, groupId: String) = api { pin(groupId, messageId, false) }

    fun removePin(messageId: String, groupId: String) = api { pin(groupId, messageId, true) }

    fun getGroupMessages(groupId: String) = api { getGroupMessages(groupId) }

    fun getGroupMessage(groupMessageId: String) = api { getGroupMessage(groupMessageId) }

    fun setGroupActionPhoto(groupActionId: String, photo: String) = api { setGroupActionPhoto(groupActionId, photo) }

    fun getGroupMember(groupId: String) = api { getGroupMember(groupId) }

    fun updateGroupMember(groupId: String, muted: Boolean, subscribed: Boolean) = api { updateGroupMember(groupId, muted, subscribed) }

    fun getGroupForPhone(phoneId: String) = api { getGroupForPhone(phoneId) }

    fun getGroupForGroupMessage(groupMessageId: String) = api { getGroupForGroupMessage(groupMessageId) }

    fun getMessagesForPhone(phoneId: String) = api { getMessagesForPhone(phoneId) }

    fun getGroupContactsForPhone(phoneId: String) = api { getGroupContactsFromPhone(phoneId) }

    fun cancelInvite(groupId: String, groupInviteId: String) = api { cancelInvite(groupId, groupInviteId) }

    fun createQuest(name: String, isPublic: Boolean, latLng: LatLng, flow: QuestFlow) = api { createQuest(QuestResult().also {
        it.name = name
        it.isPublic = isPublic
        it.geo = listOf(latLng.latitude, latLng.longitude)
        it.flow = flow
    }) }

    fun getQuests(latLng: LatLng) = api { getQuests(on<LatLngStr>().from(latLng)) }

    fun getQuest(questId: String) = api { getQuest(questId) }

    fun getQuestProgresses(questId: String) = api { getQuestProgresses(questId) }

    fun getQuestActions(questId: String) = api { getQuestActions(questId) }

    fun createQuestProgress(questId: String, ofId: String, isPublic: Boolean, active: Boolean, progress: QuestProgressFlow) = api { createQuestProgress(QuestProgressResult().also {
        it.questId = questId
        it.ofId = ofId
        it.isPublic = isPublic
        it.progress = progress
        it.active = active
    }) }

    fun updateQuestProgress(questProgressId: String, finished: Date? = null, stopped: Date? = null, active: Boolean? = null, progress: QuestProgressFlow? = null) = api {
        updateQuestProgress(questProgressId, QuestProgressResult().also {
            it.finished = finished
            it.stopped = stopped
            it.active = active
            it.progress = progress
        })
    }

    fun getQuestProgresses() = api { getQuestProgresses() }

    fun addQuestLink(questId: String, toQuestId: String) = api { addQuestLink(questId, toQuestId) }

    fun removeQuestLink(questId: String, toQuestId: String) = api { removeQuestLink(questId, toQuestId) }

    fun getQuestLinks(questId: String) = api { getQuestLinks(questId) }

    fun getQuestProgress(questProgressId: String) = api { getQuestProgress(questProgressId) }

    fun createEvent(name: String, about: String, isPublic: Boolean, latLng: LatLng, startsAt: Date, endsAt: Date, allDay: Boolean) = api { createEvent(name, about, isPublic, on<LatLngStr>().from(latLng), on<HttpEncode>().encode(on<DateFormatter>().format(startsAt))!!, on<HttpEncode>().encode(on<DateFormatter>().format(endsAt))!!, allDay) }

    fun getEvents(latLng: LatLng) = api { getEvents(on<LatLngStr>().from(latLng)) }

    fun getEvent(eventId: String) = api { getEvent(eventId) }

    fun getEventRemindersOnDay(date: Date) = api { getEventRemindersOnDay(on<HttpEncode>().encode(on<DateFormatter>().format(date))!!) }

    fun cancelEvent(eventId: String) = api { cancelEvent(eventId, true) }

    fun updateEventReminders(eventId: String, reminders: List<EventReminder>) = api { updateEventReminders(eventId, reminders) }

    fun getPhysicalGroups(latLng: LatLng) = api { getGroups("physical", on<LatLngStr>().from(latLng)) }

    fun getInactiveGroups(latLng: LatLng) = api { getGroups("inactive", on<LatLngStr>().from(latLng)) }

    fun getDirectGroups() = api { getGroups("direct") }

    fun uploadPhoto(photo: InputStream): Single<String> {
        val body = object : RequestBody() {
            override fun contentType(): MediaType? {
                return MediaType.parse("image/*")
            }

            override fun contentLength(): Long {
                return try {
                    photo.available().toLong()
                } catch (e: IOException) {
                    e.printStackTrace()
                    0
                }
            }

            override fun writeTo(sink: BufferedSink) {
                val source = Okio.source(photo)
                try {
                    sink.writeAll(source)
                } catch (e: IOException) {
                    e.printStackTrace()
                } finally {
                    Util.closeQuietly(source)
                }
            }
        }
        val part = MultipartBody.Part.createFormData("photo", "closer-photo", body)
        val id = on<Val>().rndId()
        return uiThread(api.photoUploadBackend.uploadPhoto(id, part))
                .map { id }
    }

    fun getRecentlyActivePhones(limit: Int) = api { getRecentlyActivePhones(limit) }

    fun getRecentlyActiveGroups(limit: Int) = api { getRecentlyActiveGroups(limit) }

    fun getFeatureRequests() = api { getFeatureRequests() }

    fun addFeatureRequest(name: String, description: String) = api { addFeatureRequest(name, description) }

    fun voteForFeatureRequest(featureRequestId: String, vote: Boolean) = api { voteForFeatureRequest(featureRequestId, vote) }

    fun completeFeatureRequest(featureRequestId: String, completed: Boolean) = api { completeFeatureRequest(featureRequestId, completed) }

    fun privacy() = uiThread(api.contentBackend.privacy()).map { it.string() }

    fun terms() = uiThread(api.contentBackend.terms()).map { it.string() }

    fun getPlaces(query: String, bias: LatLng, limit: Int = 3) = uiThread(api.placesBackend.query(query, bias.latitude, bias.longitude, limit))

    fun reverseGeocode(latLng: LatLng, limit: Int = 3) = uiThread(api.placesBackend.reverse(latLng.latitude, latLng.longitude, limit))

    fun getInviteCode(code: String) = api { getInviteCode(code) }

    fun createInviteCode(groupId: String, singleUse: Boolean = true) = api { createInviteCode(groupId, singleUse) }

    fun useInviteCode(code: String) = api { useInviteCode(code) }

    private fun <T> api(call: Backend.() -> Single<T>) = uiThread(call(api.backend))

    private fun <T> uiThread(observable: Single<T>): Single<T> {
        return observable.observeOn(AndroidSchedulers.mainThread()).onErrorResumeNext {
            Single.error(ApiError(it, when (it) {
                is HttpException -> it.response().errorBody()?.string()?.let { on<JsonHandler>().from(it, SuccessResult::class.java) }
                else -> SuccessResult().apply {
                    success = false
                    error = it.message
                }
            }))
        }
    }
}

class ApiError(error: Throwable, val successResult: SuccessResult?) : RuntimeException(error) {
    val error get() = successResult?.error
    val success get() = successResult?.success
}